ambry-protocol/src/main/java/com/github/ambry/protocol/ServerRequestResponseHelper.java

```diff
@@ -0,0 +1,130 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.github.ambry.protocol;
+
+import com.github.ambry.clustermap.ClusterMap;
+import com.github.ambry.network.LocalRequestResponseChannel;
+import com.github.ambry.network.NetworkRequest;
+import com.github.ambry.replication.FindTokenHelper;
+import com.github.ambry.server.ServerErrorCode;
+import io.netty.buffer.ByteBufInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+
+
+/**
+ * Helper class to decode requests coming to Ambry server
+ */
+public class ServerRequestResponseHelper {
+
+  private final ClusterMap clusterMap;
+  private final FindTokenHelper findTokenHelper;
+
+  public ServerRequestResponseHelper(ClusterMap clusterMap, FindTokenHelper findTokenHelper) {
+    this.clusterMap = clusterMap;
+    this.findTokenHelper = findTokenHelper;
+  }
+
+  /**
+   * Get deserialized request
+   * @param networkRequest incoming network request
+   * @return deserialized request
+   */
+  public RequestOrResponse getDecodedRequest(NetworkRequest networkRequest) throws IOException {
+    RequestOrResponse request;
+    if (!(networkRequest instanceof LocalRequestResponseChannel.LocalChannelRequest)) {
+      InputStream is = networkRequest.getInputStream();
+      DataInputStream dis = is instanceof DataInputStream ? (DataInputStream) is : new DataInputStream(is);
+      RequestOrResponseType requestType = RequestOrResponseType.values()[dis.readShort()];
+      switch (requestType) {
+        case PutRequest:
+          request = PutRequest.readFrom(dis, clusterMap);
+          break;
+        case GetRequest:
+          request = GetRequest.readFrom(dis, clusterMap);
+          break;
+        case DeleteRequest:
+          request = DeleteRequest.readFrom(dis, clusterMap);
+          break;
+        case TtlUpdateRequest:
+          request = TtlUpdateRequest.readFrom(dis, clusterMap);
+          break;
+        case UndeleteRequest:
+          request = UndeleteRequest.readFrom(dis, clusterMap);
+          break;
+        case ReplicaMetadataRequest:
+          request = ReplicaMetadataRequest.readFrom(dis, clusterMap, findTokenHelper);
+          break;
+        case AdminRequest:
+          request = AdminRequest.readFrom(dis, clusterMap);
+          break;
+        default:
+          throw new UnsupportedOperationException("Request type not supported");
+      }
+    } else {
+      request = (RequestOrResponse) ((LocalRequestResponseChannel.LocalChannelRequest) networkRequest).getRequestInfo()
+          .getRequest();
+      if (request.getRequestType() == RequestOrResponseType.PutRequest) {
+        // For PutRequest sent via local channel, we have to create a new PutRequest object to wrap the data buffer as a
+        // stream. Also, we need to set the crc value to null since the request is not coming over network.
+        PutRequest receivedPutRequest = ((PutRequest) request);
```

Comment:
For the PutBlob local request, getDecodedRequest will return the newly created PutRequest. 
When we send out the response for that request, we'll use this internal created PutRequest as the reference.
But it's not a LocalChannelRequest, right? Do we have problem for the sendResponse as below?
In the old implementation, although we also create the new PutBlob, but we send out the response based on the original request.

```
    @Override
  public void sendResponse(Send payloadToSend, NetworkRequest originalRequest, ServerNetworkResponseMetrics metrics) {
    LocalChannelRequest localRequest = (LocalChannelRequest) originalRequest;
    ResponseInfo responseInfo =
        new ResponseInfo(localRequest.requestInfo, null, localRequest.requestInfo.getReplicaId().getDataNodeId(),
            payloadToSend);
    BatchBlockingQueue<ResponseInfo> responseQueue = getResponseQueue(localRequest.processorId);
    responseQueue.put(responseInfo);
    logger.debug("Added response for {}, size now {}", localRequest.processorId, responseQueue.size());
  }

```

Reply:
Hmm, actually PutRequest is present inside `NetworkRequest` as `networkRequest.requestInfo.request`. In both old and new code, we are using original `NetworkRequest` in the `sendResponse()`. I think creating a new `PutRequest` should be okay.

Reply:
I see. With the new code, we still send out the response with the NetworkRequest. I missed that.